### PR TITLE
Add "Before" plugins that can cancel builds

### PR DESCRIPTION
**Pull request by siddharthist**
Fixes https://github.com/drone/drone/issues/1443

Passes automated testing, but I'm not sure how to test this on a real drone server.

### CODE REVIEW

File: yaml/parse_test.go [hunk 1]
```diff
@@ -130,6 +130,13 @@ func TestParse(t *testing.T) {
 }
 
 var sample = `
+before:
+  auth:
+    authorized_users:
+      - user1
+  when:
```
Comment (tboerger) [inline yaml/parse_test.go]: Wrong indention, this should be part of the `auth` plugin.

**Comment by siddharthist** [comment]
@tboerger Thanks for pointing that out! It's fixed, and I rebased on top of current master.

**Comment by tboerger** [comment]
LGTM

**Comment by tboerger** [comment]
Maybe we can also add some step like`validate`? :)
Maybe that makes sense for plugins like codeclimate, rubocop and other linting tools that can skip the build early :)

**Comment by siddharthist** [comment]
@tboerger Since this is just a generic slice of plugins that run before the build starts, it could hold any kind of plugin with any purpose. My own use case is build authorization, but early linting is a great one as well.

**Comment by tboerger** [comment]
@siddharthist but for linting it have to happen after cloning the repo :)

**Comment by siddharthist** [comment]
@tboerger I guess that's true hah :) what I'd love to see eventually is plugins that can occur between any step of the process. Let's open some issues for discussion after this is merged!

**Comment by tboerger** [comment]
> Let's open some issues for discussion after this is merged!

The best way yeah

**Comment by bradrydzewski** [comment]
I think as-is the before steps won't get executed unless the logic is added here:
https://github.com/drone/drone-exec/blob/master/exec/exec.go#L216:L243

If the checksum is invalid we should also prevent the before step from executing:
https://github.com/drone/drone-exec/blob/master/exec/exec.go#L96

I would also recommend testing this with `drone exec` to ensure this works as expected, after the above changes are made, to verify we didn't miss anything.

**Comment by siddharthist** [comment]
After building this with `sudo docker build --no-cache -t drone/drone-exec .`, I get the following panic:
```
drone-exec ➤ sudo drone exec             

[DRONE] starting job #1
ERRO[0000] Error starting drone/drone-exec:latest. 500 Internal Server Error: Container command not found or does not exist.

 
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4070af]

goroutine 1 [running]:
main.run(0x7f992c519048, 0xc8200c2720, 0xc820108180, 0x6, 0x8, 0xc820104600, 0x5eb, 0x0, 0x0, 0x0)
	/drone/src/github.com/drone/drone-cli/drone/exec.go:252 +0x33f
main.execCmd(0xc82000cfc0, 0x0, 0x0)
	/drone/src/github.com/drone/drone-cli/drone/exec.go:202 +0x18f9
main.glob.func8(0xc82000cfc0)
	/drone/src/github.com/drone/drone-cli/drone/exec.go:25 +0x28
github.com/drone/drone-cli/vendor/github.com/codegangsta/cli.Command.Run(0x8e86d0, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x9359d0, 0x16, 0x0, ...)
	/drone/src/github.com/drone/drone-cli/vendor/github.com/codegangsta/cli/command.go:131 +0x1055
github.com/drone/drone-cli/vendor/github.com/codegangsta/cli.(*App).Run(0xc82000cd80, 0xc82000a120, 0x2, 0x2, 0x0, 0x0)
	/drone/src/github.com/drone/drone-cli/vendor/github.com/codegangsta/cli/app.go:175 +0xfd6
main.main()
	/drone/src/github.com/drone/drone-cli/drone/main.go:44 +0x48f

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 22 [IO wait]:
net.runtime_pollWait(0x7f992c51a280, 0x72, 0xc8200101a0)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc8200dab50, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8200dab50, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc8200daaf0, 0xc820144000, 0x1000, 0x1000, 0x0, 0x7f992dd5c050, 0xc8200101a0)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc8200dc670, 0xc820144000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
net/http.noteEOFReader.Read(0x7f992c51a340, 0xc8200dc670, 0xc820114268, 0xc820144000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc8200d5a00, 0xc820144000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc8200e7680)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc8200e7680, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc820114210)
	/usr/local/go/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:685 +0xc78

goroutine 23 [select]:
net/http.(*persistConn).writeLoop(0xc820114210)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d
``` 
Any ideas?

**Comment by bradrydzewski** [comment]
seems like it is having trouble hitting the docker remote API. Do you have `DOCKER_HOST` and other docker variables set to the correct values?

**Comment by siddharthist** [comment]
@bradrydzewski No, is there some documentation you could point me to around that? I seem to remember seeing something about running drone as a Docker container without the command line tools, but I can't find it now... Thanks for the quick feedback!

**Comment by bradrydzewski** [comment]
@siddharthist sure, if you run `drone help exec` you can see a list of environment variables that drone may need in order to connect to the docker client. These are the same variables you get when you start boot2docker and then run `eval "$(boot2docker shellinit)"`

```
$ drone help exec
NAME:
   drone exec - executes a local build

USAGE:
   drone exec [command options] [arguments...]

OPTIONS:
   --docker-host "unix:///var/run/docker.sock"	docker deamon address [$DOCKER_HOST]
   --docker-tls-verify				docker daemon supports tlsverify [$DOCKER_TLS_VERIFY]
   --docker-cert-path 				docker certificate directory [$DOCKER_CERT_PATH]
```

**Comment by siddharthist** [comment]
Update: still having trouble testing, but it doesn't look related to my changes. If anybody could help run tests, that'd be awesome. Something like
```
before:
  foo: bar
  image: tianon/true
```
in a `.drone.yml` should be sufficient.

**Comment by wrouesnel** [comment]
I'd like to help test this as it implements some functionality I very much need (the ability to have plugins which can mutate the cloned source code before the build / publish steps).

**Comment by tboerger** [comment]
> I'd like to help test this as it implements some functionality I very much need (the ability to have plugins which can mutate the cloned source code before the build / publish steps).

The structure of the yaml for the upcoming version 0.5 will totally change.

**Comment by siddharthist** [comment]
@tboerger So what should happen with this PR? Just delete it?

**Comment by tboerger** [comment]
> @tboerger So what should happen with this PR? Just delete it?

I don't know about all plans of @bradrydzewski ;)

**Pull request closed without merging**